### PR TITLE
[GG] fix(spec-decode): centralize probabilistic draft sampling

### LINKED ISSUE #205: [correctness] MTP probabilistic draft Gumbel noise overlaps verifier Philox range (V2 runner)
Provenance: found during a source review of the v20 r12 release image `voipmonitor/vllm:gilded-gnosis-v20-vllmb46c3aa-si35aebc6-fi801d57a-cu132-20260730-r12`; code refs are against the composed r12 tree (`r12-composed`). Permalinks point at the last commit touching each file (blobs are byte-identical to the composed tree): speculators @ `467582afc`, `spec_decode/utils.py` @ `d179dc837`, `rejection_sampler_utils.py` / `sample/gumbel.py` @ `b453c659b`.

Deployment context: GLM-5.2 753B MoE, MTP draft, TP4, V≈154k, per-request seeds. Affected configuration: **V2 runner (`VLLM_USE_V2_MODEL_RUNNER=1`) + `draft_sample_method="probabilistic"` + sampled requests (temperature > 0)** — i.e. the release-gate MTP-probabilistic path. Greedy drafts and the V1 runner are unaffected (different code). This is a correctness/accuracy concern, not memory; proposal from a static review, happy to be corrected.

## Summary

`AutoRegressiveSpeculator.sample_draft` (inherited by `MTPSpeculator`, `EagleSpeculator`, `Gemma4Speculator`) keys the draft-proposal Gumbel noise with `positions + 1`, while the base class and every other speculator use the salted `draft_gumbel_pos(positions)` (= `positions + 1 + 2**30`) precisely so the draft stream stays disjoint from the rejection sampler's Philox range. With the unsalted override, the draft proposal noise and the verifier's acceptance/recovery/bonus noise are drawn from the **same (seed, offset) Philox pairs** every decode iteration, which breaks the independence assumptions of Leviathan-style rejection sampling.

## Where

**The override** — [`vllm/v1/worker/gpu/spec_decode/autoregressive/speculator.py#L370-L387`](https://github.com/local-inference-lab/vllm/blob/467582afcaff9ff12b44fd03240cf8b400eba489/vllm/v1/worker/gpu/spec_decode/autoregressive/speculator.py#L370-L387):

```python
logits = self.model.compute_logits(hidden_states)
if draft_logits is not None:
    # NOTE(woosuk): We must add 1 to the positions to match the Gumbel noise
    # used for draft and target sampling.
    return gumbel_sample(
        logits,
        idx_mapping,
        temperature,
        seeds,
        positions + 1,
        ...
```

**The documented contract it bypasses** — [`vllm/v1/worker/gpu/spec_decode/utils.py#L10-L23`](https://github.com/local-inference-lab/vllm/blob/d179dc83755ca7365a6c1b1294c74d7908106bc7/vllm/v1/worker/gpu/spec_decode/utils.py#L10-L23):

```python
# Salt added to the Philox offsets used for draft-token Gumbel noise.
# Positions are bounded by max_model_len, so this puts the draft stream in a
# range disjoint from the target-side offsets.
DRAFT_GUMBEL_POS_OFFSET = 1 << 30


def draft_gumbel_pos(positions: torch.Tensor) -> torch.Tensor:
    """Philox offsets for draft Gumbel noise, given draft-row positions.

    The rejection sampler keys both its acceptance uniform and recovery
    Gumbel noise for position P by offset P. Keep the draft proposal on a
    disjoint Philox range so rejection sampling remains unbiased.
    """
    return positions + (1 + DRAFT_GUMBEL_POS_OFFSET)
```

The base class uses it ([`speculator.py#L366-L379`](https://github.com/local-inference-lab/vllm/blob/467582afcaff9ff12b44fd03240cf8b400eba489/vllm/v1/worker/gpu/spec_decode/speculator.py#L366-L379): `gumbel_sample(..., draft_gumbel_pos(positions), ...)`), and DFlash/DSpark go out of their way to honor it (`dflash/speculator.py:461-463`, `dspark/speculator.py:255-265`: "…so pass Q-2 to get a key unique per predicted position **and disjoint from the rejection sampler's acceptance/recovery keys**"). Only the autoregressive override drops the salt.

## Mechanism — the streams do collide

**Same seed tensor.** Both sides use the identical per-request seeds: the runner passes `self.sampler.sampling_states.seeds.gpu` into `speculator.propose` (`model_runner.py:2032-2033`), the speculator copies it (`speculator.py:397-398`), and the rejection sampler receives the same tensor (`gpu/spec_decode/rejection_sampler.py:148-149`). Both index it by the same `req_state_idx`.

**Same offset range.** The V2 rejection sampler keys its randomness by the **absolute row position** (`pos = input_batch.positions[logits_indices]`, `gpu/spec_decode/rejection_sampler.py:130` → `rejection_sample(..., pos, ...)` at `:144`):

- acceptance uniform for the i-th draft token, [`rejection_sampler_utils.py#L530-L534`](https://github.com/local-inference-lab/vllm/blob/b453c659b7218b77690cbc3e343814c9b1d98780/vllm/v1/worker/gpu/spec_decode/rejection_sampler_utils.py#L530-L534):

```python
pos = tl.load(pos_ptr + logit_idx)
u = tl_rand32(seed, pos, includes_zero=False)
```

- recovery/bonus Gumbel, `_resample_kernel` → `gumbel_block_argmax` with the same `pos_ptr` (`rejection_sampler_utils.py:776-793`), which derives ([`gumbel.py#L147-L158`](https://github.com/local-inference-lab/vllm/blob/b453c659b7218b77690cbc3e343814c9b1d98780/vllm/v1/worker/gpu/sample/gumbel.py#L147-L158)):

```python
seed = tl.load(seeds_ptr + req_state_idx, ...)
pos = tl.load(pos_ptr + token_idx)
gumbel_seed = tl.randint(seed, pos)
...
u = tl_rand32(gumbel_seed, block, includes_zero=False)
```

Let the verify batch for a request be rows `[t@R, d1@R+1, …, dk@R+k]` (R = position of the last committed token). The verifier consumes offsets `{R, …, R+k}`: acceptance `u` for `d_{i}` at offset `R+i-1`; recovery Gumbel after `a ≥ 1` acceptances at offset `R+a`; bonus Gumbel at offset `R+k`. The draft, meanwhile, proposed `d_j` (which sits at position `R+j`) with key `positions+1 = R+j` — inside the same `{R+1, …, R+k}` window, same seed. Concretely, **every decode iteration**:

1. **Acceptance uniform ← draft seed word.** Triton's `tl.rand(seed, off)` is `uint32_to_uniform_float(tl.randint(seed, off))` — the same Philox word. So the acceptance uniform for `d_{j+1}` (offset `R+j`) is the float reinterpretation of the exact 32-bit word used as `gumbel_seed` for proposing `d_j`. The uniform that decides whether to accept a token is a deterministic function of the noise that generated the previous, already-in-context draft token — the u ⟂ (draft, distributions) independence in the Leviathan et al. proof no longer holds.
2. **Recovery noise ≡ draft proposal noise.** On rejection after `a ≥ 1` acceptances, the recovery Gumbel field is `tl_rand32(tl.randint(seed, R+a), v)` over vocab `v` — bit-identical to the field the draft used to propose `d_a` (same construction, same global vocab indexing on both sides). The residual `max(p−q, 0)` resample is performed with the same noise realization that selected the last accepted token.
3. **Bonus noise ≡ last draft's proposal noise.** When all k drafts are accepted, the bonus token at offset `R+k` reuses `d_k`'s exact Gumbel field: bonus = `argmax(p/T + g)` where `d_k = argmax(q/T + g)` with the *same* g. When the target and draft distributions are close (they are, for a good MTP head — that's the point), this correlates the bonus token with the preceding draft token far beyond what independent sampling would produce (a plausible symptom is repetition amplification in sampled output).

Note the override's comment says the +1 is "to match the Gumbel noise used for draft and target sampling", but no valid coupling is achieved either: the target-side convention keys the token at position Q by offset Q−1 (main sampler `pos = positions[logits_indices]`; verifier as above), while the override keys it by Q. The result is not a coupled-sampling scheme — it's an off-by-one collision with the verifier's keys for neighboring positions.

## Suggested fix

One line — delegate to the documented contract in the override:

```python
return gumbel_sample(
    logits, idx_mapping, temperature, seeds,
    draft_gumbel_pos(positions),   # was: positions + 1
    ...
)
```

(or simply drop the override's probabilistic branch and call `super().sample_draft(...)`, keeping the greedy short-circuit). `draft_gumbel_pos` keeps keys unique per predicted position and disjoint from the verifier by construction. History suggests this is an incomplete migration: the salt was introduced in `d179dc837` in the base class, and the autoregressive override kept the pre-salt `positions + 1`.

## Verification hint

- Instrument both sides and log `(seed, offset)` pairs for one sampled request with MTP probabilistic on the V2 runner: the draft's `gumbel_sample` offsets equal the verifier's acceptance/recovery offsets shifted by one position, in the same iteration.
- Statistical A/B: fixed seed, temperature 1, compare the empirical distribution of bonus tokens (and rejected-position recovery tokens) against pure target sampling, before/after the one-line fix; the unsalted variant shows excess `P(bonus == d_k)`.

Cross-ref: memory-side findings for the same path are in #200.

### PR DESCRIPTION
**Pull request by voipmonitor**
## Summary

Centralize probabilistic draft Gumbel sampling in `DraftModelSpeculator._sample_probabilistic_draft` and route the generic, autoregressive, and DSpark paths through it.

The shared method always applies `draft_gumbel_pos()`, keeping proposal noise disjoint from verifier acceptance/recovery offsets. The autoregressive caller still forwards `active_num_reqs`, preserving its FULL-graph padded-row safeguard.

Fixes #205.
Supersedes #209.

## Why this replaces #209

#209 correctly fixes the live GG autoregressive override, but only changes that call site. GG already contains the broader salted-stream backport corresponding to upstream vLLM #47386; a later GG autoregressive override copied the old upstream `positions + 1` call and bypassed the corrected base implementation.

Keeping raw `gumbel_sample()` calls in individual speculators makes that failure mode easy to repeat. This PR makes stream ownership a base-speculator responsibility instead:

- generic probabilistic drafts use the shared method;
- the GG autoregressive override uses it while retaining `active_num_reqs`;
- DSpark sequential probabilistic sampling uses the same method;
- DFlash continues to inherit the generic path.

Only the base method calls `gumbel_sample()` in `spec_decode`; individual speculators pass model positions, not Philox offsets.

## Correctness scope

Affected before this fix:

- V2 autoregressive speculators;
- probabilistic draft sampling;
- sampled target requests.

Unaffected:

- greedy drafts and temperature-zero requests;
- the V1 proposer;
- the all-accepted bonus draw.

The first-order defect is rejection recovery reusing the exact vocabulary-sized Gumbel field that selected the rejected draft token. Immediate acceptance probability is unchanged; the fix restores the recovery sampler's independence contract.

## Runtime invariants

- No new allocation, kernel launch, synchronization, or public API.
- The same single tensor add is used to derive `positions + 1 + 2**30`.
- AR FULL-graph active-row protection is unchanged.
- DSpark's existing predicted-position convention is unchanged.
- Probabilistic outputs can change bit-for-bit as expected from correcting the RNG stream.

## Verification

Release-container tests:

```text
4 passed
```

- AR tuple/tensor behavior
- shared salted-offset contract
- AR forwarding of the FULL-graph active-row limit

GPU tests on CUDA:

```text
11 passed
```

- probabilistic rejection output-distribution regression
- Gumbel sampler tests

Static checks:

```text
Ruff: pass
format: pass
git diff --check: pass
```

The existing statistical regression inherited from upstream #47386 measures the coupled stream outside its target-distribution noise floor and the salted stream within it; this PR preserves that implementation and closes the GG override gap without claiming an unmeasured model-level quality delta.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved probabilistic draft-token sampling across autoregressive and DSpark decoding.
  * Preserved sampling positions, random seeds, temperature, and active-request state for more consistent results.
  * Added support for target-vocabulary mappings and safer handling of processed logits.

* **Tests**
  * Added coverage for sampler arguments, disjoint random-sampling positions, and state preservation during autoregressive sampling.

### CODE REVIEW

**Comment by malaiwah** [comment]
## Rental-machine validation — PASS (focused and composed-source gates)

Tested merged commit: `30038602b71395f481ef4a6edfe4fcf8551d9c15`  
Original PR head: `7bf5ddd98ffc1d503ad6017e9a0e8edd4bc9bf0d`  
Tested branch: `dev/gilded-gnosis`  
Composed stack: #211 + #210 at `b135a9b79377386ee805c22dee950e985f07ed28`

Host: Vast.ai 46335896, Ubuntu 24.04.4 / kernel 7.0.0-28, 2x EPYC 9455, 1.5 TiB RAM.  
GPU/topology: 4x RTX PRO 6000 Blackwell; all `NODE` on NUMA 0; all 12 directed P2P copies passed.  
Driver/CUDA/PyTorch: 610.43.02 / CUDA UMD 13.3 / PyTorch 2.12.0+cu132 (CUDA 13.2).  
Test environment: retained turnkey container; isolated cold/warm caches.

### Commands and results

- `pytest tests/v1/worker/test_gpu_autoregressive_speculator.py` — 4 passed; Ruff clean.
- GPU Gumbel/rejection-sampler suites — 11 passed cold (19.75 s), 11 passed warm (16.84 s).
- Focused DSpark spy — PASS; caller supplied exactly `sample_pos[:, i] - 2`.
- With #210 — 16 combined CPU tests passed; #211 GPU suites passed 11 cold and 11 warm.

Greedy/temperature-zero coverage passed and no capture-time compile, Xid, hang, or request error was observed. The later full-model live-probabilistic gate was not run because independent SparkInfer and LMCache source-composition gates failed first.

Report and durable logs: [field-review result](https://github.com/malaiwah/glm52-exl3-vast/blob/codex/field-review-pr-validation/docs/field-review-results/2026-07-30-vast-46335896/README.md#211--probabilistic-draft-sampling)

No unreported source modifications were present; the tested worktrees were clean.

---

### Exact r14-base composition — PASS

The immutable r14 vLLM lock was replayed from base `f978d009fab996617f9a3cadef36ce727bcd83cd` and independently matched published tree `749050edab1b6664937c52fa1b0be360be632c1e`. Original #211 head `7bf5ddd98ffc1d503ad6017e9a0e8edd4bc9bf0d` plus exact #210 then applied without conflict.

Resulting candidate `5a7ac1481e1fc24b0bbc35efe160b2ff34797bee` passed **16/16 CPU/static tests** and the #211 probabilistic GPU union **11/11 cold and 11/11 warm** on physical GPU 0. Ruff and `diff --check` were clean; unchanged r14 native binaries used for worktree import were hash-recorded and removed after the gate.

This extends the earlier current-development-branch result to the actual r14 image source boundary. Exact evidence: [r14 ledger at `4d2aa18`](https://github.com/malaiwah/glm52-exl3-vast/blob/4d2aa18/docs/field-review-results/2026-07-30-vast-46335896/UPSTREAM-REPAIR-CAMPAIGN.md#f-exact-r14-base-reconstruction).

**Comment by malaiwah** [comment]
## Live full-model probabilistic follow-up — PASS

Tested merged change: `30038602b71395f481ef4a6edfe4fcf8551d9c15` (original PR head `7bf5ddd98ffc1d503ad6017e9a0e8edd4bc9bf0d`)  
Exact r14 composed vLLM candidate: `5a7ac1481e1fc24b0bbc35efe160b2ff34797bee`  
Host: Vast.ai 46335896, 4x RTX PRO 6000 Blackwell, driver 610.43.02 / CUDA UMD 13.3 / PyTorch 2.12.0+cu132.

The previously missing live gate is now complete. The full GLM-5.2 3.25-bpw EXL3 server ran the V2 runner with MTP3, `draft_sample_method=probabilistic`, temperature 1.0, and a fixed seed. This does execute #211: `method=mtp` selects `MTPSpeculator`, which inherits `AutoRegressiveSpeculator`; non-null probabilistic draft logits delegate to the shared salted sampler.

Observed across 32 live requests / 4,096 generated tokens:

- 1,526 speculative steps;
- 4,578 drafted tokens;
- 2,563 accepted tokens;
- pooled mean acceptance length: **2.680**;
- pooled draft-token acceptance: **55.99%**;
- **0 request errors, 0 preemptions**;
- all short/32K/65K/126K correctness and retrieval checks passed without degeneration.

The exact-r14 focused distribution suite already passed **11/11 cold and 11/11 warm**, including greedy and temperature-zero controls. A separate plain-checkpoint download is unnecessary: this EXL3 MTP path reaches the same changed sampler.

Log scope: cold priming compiled 28 previously unseen SparkInfer request shapes, which the test plan permits. The subsequent warm audit contains **zero** compile, ERROR, CUDA/IPC, or process-failure findings. Four XGrammar FSM errors occurred earlier in structured-output checks because this exact-source gate deliberately disabled the appliance compatibility patch to preserve SHA attribution; that separate built-image requirement is not attributable to #211.

Evidence: [campaign ledger](https://github.com/malaiwah/glm52-exl3-vast/blob/3f48f69/docs/field-review-results/2026-07-30-vast-46335896/UPSTREAM-REPAIR-CAMPAIGN.md#i-full-model-pre-composition-control), [raw benchmark JSON](https://github.com/malaiwah/glm52-exl3-vast/blob/3f48f69/docs/field-review-results/2026-07-30-vast-46335896/artifacts/vllm-r14-field-baseline-mtp3-c8-v2-benchmark.json), and [clean warm-window audit](https://github.com/malaiwah/glm52-exl3-vast/blob/3f48f69/docs/field-review-results/2026-07-30-vast-46335896/artifacts/vllm-r14-field-baseline-mtp3-c8-v2-warm-log-audit.json).
